src/ui/carousel/carousel-synchronizer.ts

```diff
@@ -0,0 +1,173 @@
+ *   Array.from(document.querySelectorAll('.container-a .slide')),
+ *   {
+ *     allowLooping: true,
+ *     transition: new DraggableSlide(),
+ *   });
+ * const b = new Carousel(
+ *   document.querySelector('.container-b'),
+ *   Array.from(document.querySelectorAll('.container-b .slide')),
+ *   {
+ *     allowLooping: true,
+ *   });
+ * new CarouselSynchronizer(a, b);
+ *
+ * // At some point later, dispose.
+ * a.dispose();
+ * b.dispose();
+ * syncInstance.dispose(a);
+ * ```
+ *
+ * CSS:
+ * ```
+ * .carousel {
+ *   position: relative;
+ *   height: 20vh;
+ *   width: 80vw;
+ *   overflow: hidden;
+ *   display: flex;
+ *   flex-wrap: nowrap;
+ * }
+ * .slide {
+ *   display: flex;
+ *   justify-content: center;
+ *   align-items: center;
+ *   height: 100%;
+ *   width: 50%;
+ *   flex-shrink: 0;
+ *   font-size: 10vw;
+ *   user-select: none;
+ * }
+ * .slide:nth-child(1) {
+ *   background: paleturquoise;
+ * }
+ * .slide:nth-child(2) {
+ *   background: palevioletred;
+ * }
+ * .slide:nth-child(3) {
+ *   background: palegreen;
+ * }
+ * .nav {
+ *   display: flex;
+ *   justify-content: center;
+ *   align-items: center;
+ * }
+ * .dot {
+ *   flex-grow: 0;
+ *   flex-shrink: 0;
+ *   width: 10px;
+ *   height: 10px;
+ *   border-radius: 50%;
+ *   opacity: .25;
+ *   background: black;
+ *   transition: opacity 1s, transform 1s;
+ *   margin: 0 8px;
+ * }
+ * .dot.active {
+ *   transform: scale(1.25);
+ *   opacity: 1;
+ * }
+ * .dot.before, .dot.after {
+ *   transform: scale(0.9);
+ * }
+ * ```
+ */
+export class CarouselSynchronizer {
+  private graph: DefaultMap<Carousel, Set<Carousel>>;
+
+  constructor(...carousels: Carousel[]) {
+    this.graph = DefaultMap.usingFunction((carousel: Carousel) => new Set());
+    this.sync(...carousels);
+  }
+
+  /**
+   * Synchronize the given carousels.
+   */
+  sync(...carousels: Carousel[]) {
+    // Create a set with the given carousels, if any of these carousels are
+    // already synced to other carousels merge those sets together so that
+    // everything stays synchronized.
+    carousels.forEach((carousel) => {
+      if (!this.graph.has(carousel)) {
+        carousel.on(
+            CarouselEvent.TRANSITION_START,
+            () => this.handleCarouselTransition(carousel));
+      }
+
+      const set = this.graph.get(carousel);
+      carousels
+          .filter((c) => c !== carousel)
+          .forEach((c) => set.add(c));
+    });
+  }
+
+  /**
+   * Handle the transition of the given carousel to the given index.
+   *
+   * Will transition all other carousels synced with the given carousel to the
+   * given index.
+   *
+   * @param carousel
+   */
+  handleCarouselTransition(carousel: Carousel) {
+    const transitionTarget = carousel.getTransitionTarget();
+    if (!transitionTarget || transitionTarget.drivenBySync) {
+      return; // Skip if this sync was triggered by a sync call
+    }
+
+    const syncedCarousels = this.graph.get(carousel);
```

Comment:
I think you may need to walk all the linked nodes in the graph since I think the following example would fail:

```
a.sync(b)
c.sync(d)
b.sync(c)
```

^ the above would produce a graph like:
```
graph = {
  a: [b]
  b: [a, c]
  c: [b, d]
}
```

If the following is called:

```
c.transitionToIndex(2)
```

Then only `b` and `d` would be updated but not `a`.

You can sorta conceptualize this example using islands connected by bridges. Imagine you start on an island and need to send a message to every connected island once and only once by crossing bridges. You'll need to keep track of which islands have already been visited and avoid crossing bridges that lead to those islands.

Reply:
Looks like I made a mis-step somewhere in the conversion to graphs rather than sets. I'll revisit the implementation, this case was definitely handled previously so it should be again.

Reply:
Working demo pushed up. The issue I was seeing was a mistake I made while working on removing TransitionTarget, I wasn't passing a proper value through the data.

Reply:
Removed CarouselSynchronizer. Syncing is now built into Carousel and related carousels stored as sets, merged as needed. All synced carousels share the same set instance.